edgelet/edgelet-docker/src/runtime.rs

```diff
@@ -798,61 +798,56 @@ where
     Box::new(match pid {
         Pid::None => Either::A(future::ok(AuthId::None)),
         Pid::Any => Either::A(future::ok(AuthId::Any)),
-        Pid::Value(pid) => {
-            Either::B(
-                // to authenticate request we need to determine whether given pid corresponds to
-                // any pid from any module. In order to do so, we are getting a list of all running
-                // modules and for each of them execute docker top command. There can be errors
-                // during a multiple requests, so we are filtered out those modules that we active
-                // during docker inspect operation but have gone after (NotFound and TopModule errors).
-                runtime
-                    .list()
-                    .map(move |list| {
-                        stream::futures_unordered(list.into_iter().filter_map(|module| {
-                            if expected_module_id.is_none()
-                                || *expected_module_id.as_ref().unwrap() == module.name()
-                            {
-                                Some(module.top())
-                            } else {
-                                None
-                            }
-                        }))
-                    })
-                    .into_stream()
-                    .flatten()
-                    .then(|result| match result {
-                        Ok(top) => Ok(Some(top)),
-                        Err(err) => match err.kind() {
-                            ErrorKind::NotFound(_)
-                            | ErrorKind::RuntimeOperation(RuntimeOperation::TopModule(_)) => {
-                                Ok(None)
+        Pid::Value(pid) => Either::B(
+            // to authenticate request we need to determine whether given pid corresponds to
+            // any pid from a module with provided module name. In order to do so, we are
+            // load a list of all running modules and execute docker top command only for
+            // the module that have corresponding name. There can be errors during requests,
+            // so we are filtered out those modules that we active during docker inspect
+            // operation but have gone after (NotFound and TopModule errors).
+            match expected_module_id {
+                None => Either::A(future::ok(AuthId::None)),
+                Some(expected_module_id) => Either::B(
+                    runtime
+                        .list()
+                        .map(move |list| {
+                            list.into_iter()
+                                .find(|module| expected_module_id == module.name())
+                        })
+                        .and_then(|module| module.and_then(|module| Some(module.top())))
+                        .then(|result| match result {
```

Comment:
This `then` is checking for the error from `runtime.list()`, not `module.top()`, so it will never get `ErrorKind::NotFound`

Reply:
Sure